File: pt5/1117.building-h-2-o.py
from __future__ import annotations

from threading import Barrier, Semaphore
from typing import Callable


class H2O:
    def __init__(self):
        self.hydrogen_slots = Semaphore(2)
        self.oxygen_slots = Semaphore(1)
        self.molecule_barrier = Barrier(3)

    def hydrogen(self, releaseHydrogen: Callable[[], None]) -> None:
        self.hydrogen_slots.acquire()
        releaseHydrogen()
        self.molecule_barrier.wait()
        self.hydrogen_slots.release()

    def oxygen(self, releaseOxygen: Callable[[], None]) -> None:
        self.oxygen_slots.acquire()
        releaseOxygen()
        self.molecule_barrier.wait()
        self.oxygen_slots.release()

File: pt5/1117.building-h-2-o.java
import java.util.concurrent.BrokenBarrierException;
import java.util.concurrent.CyclicBarrier;
import java.util.concurrent.Semaphore;

class H2O {
    private final Semaphore hydrogenSlots = new Semaphore(2);
    private final Semaphore oxygenSlots = new Semaphore(1);
    private final CyclicBarrier moleculeBarrier = new CyclicBarrier(3);

    public H2O() {}

    public void hydrogen(Runnable releaseHydrogen) throws InterruptedException {
        hydrogenSlots.acquire();
        releaseHydrogen.run();
        awaitMolecule();
        hydrogenSlots.release();
    }

    public void oxygen(Runnable releaseOxygen) throws InterruptedException {
        oxygenSlots.acquire();
        releaseOxygen.run();
        awaitMolecule();
        oxygenSlots.release();
    }

    private void awaitMolecule() throws InterruptedException {
        try {
            moleculeBarrier.await();
        } catch (BrokenBarrierException ex) {
            throw new RuntimeException(ex);
        }
    }
}

